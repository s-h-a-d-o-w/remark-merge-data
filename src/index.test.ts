import { readFile } from "node:fs/promises";
import { join } from "node:path";
import { remark } from "remark";
import { MergeDataOptions, remarkMergeData } from "./index.js";

describe(remarkMergeData.name, () => {
  it("merges the readme example correctly", async () => {
    const options: MergeDataOptions = {
      lang: "some-language",
      data: {
        foo: "bar",
        unchanged: 123,
      },
    };

    expect(
      (
        await remark()
          .use(remarkMergeData, options)
          .process(
            await readFile(
              join(import.meta.dirname, "fixtures/readme-example.md"),
            ),
          )
      ).value,
    ).toMatchSnapshot();
  });

  it("merges JSON deeply correctly", async () => {
    const options: MergeDataOptions = {
      lang: "json",
      data: {
        globalRootProp: true,
        nested: {
          globalProperty: "global",
          globalArray: [0],
          overwrittenArray: [3, 4, 5],
        },
      },
    };

    expect(
      (
        await remark()
          .use(remarkMergeData, options)
          .process(
            await readFile(join(import.meta.dirname, "fixtures/json.md")),
          )
      ).value,
    ).toMatchSnapshot();
  });

  it("merges YAML deeply correctly", async () => {
    const options: MergeDataOptions = {
      isYaml: true,
      lang: "yaml",
      data: `
globalRootProp: true
nested:
  globalProperty: global
  globalArray:
    - 0
  overwrittenArray:
    - 3
    - 4
    - 5
`,
    };

    expect(
      (
        await remark()
          .use(remarkMergeData, options)
          .process(
            await readFile(join(import.meta.dirname, "fixtures/yaml.md")),
          )
      ).value,
    ).toMatchSnapshot();
  });

  it("throws if yaml is supposed to be used but data isn't a string", async () => {
    const options: MergeDataOptions = {
      isYaml: true,
      lang: "yaml",
      data: {},
    };

    await expect(async () =>
      remark()
        .use(remarkMergeData, options)
        .process(await readFile(join(import.meta.dirname, "fixtures/yaml.md"))),
    ).rejects.toMatchInlineSnapshot(
      `[Error: If you use YAML, you have to provide YAML code as a string.]`,
    );
  });
});
