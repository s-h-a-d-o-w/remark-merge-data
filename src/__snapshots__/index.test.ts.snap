// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`remarkMergeData > global data doesn't get polluted by merges 1`] = `
"\`\`\`json
{
  "pollution1": true
}
\`\`\`
"
`;

exports[`remarkMergeData > looks for an exact match of the meta data 1`] = `
"\`\`\`json type=match foo=bar
{
  "sharedProp": true
}
\`\`\`
"
`;

exports[`remarkMergeData > merges JSON deeply correctly 1`] = `
"\`\`\`json
{
  "globalRootProp": true,
  "nested": {
    "globalProperty": "global",
    "globalArray": [
      0
    ],
    "overwrittenArray": [
      0,
      1,
      2
    ],
    "localProperty": 123
  },
  "localRootProp": "bar"
}
\`\`\`
"
`;

exports[`remarkMergeData > merges YAML deeply correctly 1`] = `
"\`\`\`yaml
globalRootProp: true
nested:
  globalProperty: global
  globalArray:
    - 0
  overwrittenArray:
    - 0
    - 1
    - 2
  localProperty: 123
localRootProp: bar

\`\`\`
"
`;

exports[`remarkMergeData > merges the readme example correctly 1`] = `
"\`\`\`some-language
{
  "foo": "bar",
  "localProp": true
}
\`\`\`

\`\`\`some-other-language
{
  "foo": "baz",
  "localProp": true
}
\`\`\`
"
`;
